[WiP] DaemonSet updates

lukaszo opened the pull request:
This patch it built on top of #19627. Big part of it was written
by @madhusudancs.
I rebased the code on the current master and fixed unit tests.

I've done it to be as much similar as possible to deployment update.

Currently there are two update strategies:
- NoopDaemonSetStrategyType(default) - it does nothing. It's backward compatible.
- RollingUpdateDaemonSetStrategyType - it deletes current pod(s) and creates new with updated pod template. In one step it updates no more than `MaxUnavailable` pods.

I implemented `Pause` option. It can be run by `kubectl rollout pause ds/<name>`. When pause is enabled DaemonSetController will not mange this DaemonSet it will only update its status. If it will be paused during the update, update will also be paused.
To resume it user needs to run `kubectl rollout resume ds/<name>`.

To be able to rollback to the previous version I implemented update history. As @bgrant0607 suggested I used api.PodTemplate type for it. 
When daemon is created PodTemplate with version 1 is created. When daemon is updated new PodTemplate is created with version 2. Versions are stored as an annotations. 
When rollback is run:
- I'm looking for PodTemplate with a given version. 
- copy PodTemplate.Spec to daemon.Spec
- rollingUpdate sees that pods have different spec version and deletes them and creates new pods.

All `kubectl rollout` commands are working with DaemonSets.

It's still work in progress. I need to polish the code and test it with more advanced workloads.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31693)

<!-- Reviewable:end -->

madhusudancs commented:
cc @bgrant0607

bgrant0607 commented:
ok to test

Thanks @lukaszo. We're trying to get 1.4 out the door at the moment, so it may be a while before we can provide a thorough review, but I'll provide some quick comments now.

AutoUpdate: I imagined a special rollout strategy instead.

I recommend taking a look at the Deployment implementation. We encountered a number of tricky cases when implementing it, such as the need to cull unhealthy pods and changes in the pod template hash across releases.

We'll similarly want to support history and rollback for DaemonSet, though some of the details would be different. It might require use of the little-known pod template API that was created for #170.

For a variety of reasons, it would be much simpler to start with a client-side implementation: https://github.com/kubernetes/kubernetes/issues/22543#issuecomment-215833295

cc @pwittrock @erictune

lukaszo commented:
> Thanks @lukaszo. We're trying to get 1.4 out the door at the moment, so it may be a while before we can provide a thorough review, but I'll provide some quick comments now.

Big thanks @bgrant0607 for comments. I know that everyone is working on 1.4. I'm targeting 1.5 release.

> AutoUpdate: I imagined a special rollout strategy instead.

I removed `AutoUpdate` and added NoopStrategy which is default option now. It's backward compatible.

> I recommend taking a look at the Deployment implementation. We encountered a number of tricky cases when implementing it, such as the need to cull unhealthy pods and changes in the pod template hash across releases.

I believe I've read all issues related to Deployment and DaemonSet updates :D I'm aware of this two described cases and I will handle them.

> We'll similarly want to support history and rollback for DaemonSet, though some of the details would be different. It might require use of the little-known pod template API that was created for #170.

I'm currently looking into Deployment  history implementation. I hope to push changes for DaemonSet at the beginning of the next week.

> For a variety of reasons, it would be much simpler to start with a client-side implementation: #22543 (comment)

client-side implementation is a dead end. Some parts, like history will look the same but I prefer to work on server side version because it's "more enterprise" solution :)

lukaszo committed:
DaemonSet update history

lukaszo committed:
DaemonSet rollback

bgrant0607 commented:
cc @davidopp

lukaszo committed:
e2e test for daemon set updates

bgrant0607 commented:
@lukaszo Please explain what this PR does in the description. What is the update strategy? How does pause work? How does rollback work?

lukaszo commented:
@bgrant0607  I need to resolve one more issue and it will be ready for a review. I will update the description and record [asciinema](asciinema.org) demo

bgrant0607 commented:
@lukaszo I really recommend not doing more work on this until we agree on what the right approach is.

lukaszo committed:
DaemonSet updates

This patch it built on top of #19627. Big part of it was written
by @madhusudancs.
I rebased the code on the current master and fixed unit tests.
I also added option to pause update with:

kubectl rollout pause/resume daemonset/<name>

I also introduced new paramater
`Spec.AutoUpdate` which for backward compatibility is set to False.

Related issue: #15310

lukaszo commented:
@bgrant0607  I updated the description.

lukaszo committed:
Remove AutoUpdate and add Noop strategy type

lukaszo commented:
@bgrant0607 I would like to continue to work on it. Can you please comment?

lukaszo commented:
Btw @bgrant0607  @davidopp Should I create issue for it in the features repo?

lukaszo committed:
Generated part for DaemonSet updates

bgrant0607 commented:
@lukaszo 

I doubt this will get into 1.5. 

We want this feature eventually and I appreciate your work on it, but we're unable to review it now, and it will need more than the average amount of testing. It's risky and complex. Additionally, we already have more than 300 test-failure issues open, and this feature is likely to be prone to such intermittent failures.

Could you come to the next Contributor Experience meeting to discuss how we could better handle such cases?

lukaszo commented:
@bgrant0607

I understand your doubts.  We @team/mirantis can guarantee that we will maintain this functionality. 
- we have own QA team which is doing manual and automated testing 
- we have a big team, even if I stop working on it then someone else will pickup it up
- we need this feature for our `OpenStack on Kubernetes`. Some services will run as  DaemonSets so we need it to be production ready. Upgrade is one of the major requirements.
- we also run performance test so we will be able to test it on at least 400 nodes

I will of course participate in the next Contributor Experience meeting. For now I can propose that I can split this PR to smaller chunks. Rebasing it is a really PITA.

bgrant0607 commented:
@lukaszo Go ahead and create an issue in the features repo. We'll need one eventually, anyway.

idvoretskyi commented:
@bgrant0607 your concerns are reasonable, at the same time as @lukaszo has mentioned, Mirantis has enough resources to handle the full scope of developing the feature. This feature is actually needed in Kubernetes; @kubernetes/sig-openstack is going to be one of the feature consumers.

@lukaszo please, go ahead and create a feature tracking issue. Please, don't forget that submission window for 1.5 release features closes on Monday, Oct 10.

bgrant0607 commented:
@lukaszo 

As discussed:

We don't have time to carefully review the implementation in the 1.5 timeframe.

Once the main OWNER PR is merged, you'll need to add yourself as a reviewer.

Update features will need to start as alpha. Representation is TBD, pending discussion in #30819. Will follow up there.

e2e tests need to be added to the alpha e2e test suite, and we need to ensure it is working.

With this feature disabled, we might not be able to create podtemplate objects at all.

@derekwaynecarr Do Openshift clusters generally set resourcequota for podtemplate objects to 0 by default?

bgrant0607 commented:
Example e2e config for petset:
https://github.com/kubernetes/test-infra/blob/master/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-gce.yaml#L926

There are also configs that look like they are supposed to run all alpha features, but it's not clear that it's working:
https://github.com/kubernetes/test-infra/blob/master/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-gce.yaml#L201

therc commented:
I'd really love to see this implemented and cry a little on the inside whenever I need to update a DS, but, at the same time, IMHO this requires some kind of canarying, even more so than deployments (which really need it already). A botched DS update can hose a cluster, e.g. if you use it to deploy something like kube2iam.

bgrant0607 commented:
@lukaszo It looks like we're going to stay with annotations for alpha features for now. Given that, is there any reason updates couldn't be driven by an out-of-tree controller in 1.5, similar to the initial prototypes of ingress?

lukaszo commented:
@bgrant0607  we can do out-of-tree controller in 1.5, sure. The UX will be much worse though. In this PR user can use `kubectl rollout` to get status/history and to do a rollback. With out-of-tree controller user will be able to only use `kubectl edit`. Rollback will not be even possible without changes in api :(
I really would prefer not to go this way.

I would prefer to go with alpha features but to stay in the tree.
If you think that `pause` is risky, I can remove it. Then daemoncontroller.go will be changed only in two places. Rest of the new code will not be even executed if update is not used.

kargakis commented on pkg/controller/daemon/util/daemon_util.go:
```diff
@@ -0,0 +1,161 @@
+	return max
+}
+
+// Revision returns the revision number of the input replica set
```
s/replica set/daemon set/

kargakis replied:
s/replica set/pod template/

kargakis commented on pkg/controller/daemon/util/daemon_util.go:
```diff
@@ -0,0 +1,161 @@
+}
+
+// TODO: Should I make it a real controller?
+type PodTemplateControllerInterface interface {
```
Why can't you use the podTemplate namspacer directly?

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -89,6 +89,9 @@ type DaemonSetsController struct {
 	// Added as a member to the struct to allow injection for testing.
 	nodeStoreSynced cache.InformerSynced
 
+	// To allow injection of CreatePodTemplate for testing
```
I don't think you need this. You can just use `kubeClient` and fake the pod template calls, right?

lukaszo replied:
Yeah, I'm using this technique in daemon_util_test.go. I will change it here also

kargakis commented on pkg/controller/daemon/util/daemon_util.go:
```diff
@@ -0,0 +1,161 @@
+		return nil, err
+	}
+	if len(podsTemplates.Items) == 0 {
+		podTemplate, err = CreatePodTemplateFromDS(ptc, ds, "1")
```
Admittedly, it's the first time I see us using pod templates as objects. Is there any related discussion somewhere about it?

lukaszo replied:
Only this comment from @bgrant0607  https://github.com/kubernetes/kubernetes/pull/31693#issuecomment-243675435

I have to say that using it was quite pleasant. Do you see other options to store history?

kargakis commented on pkg/controller/daemon/util/daemon_util.go:
```diff
@@ -0,0 +1,161 @@
+		daemonTemplateHash := podutil.GetPodTemplateSpecHash(ds.Spec.Template)
+		daemonTemplateHashStr := strconv.FormatUint(uint64(daemonTemplateHash), 10)
+		for _, podTmpl := range podsTemplates.Items {
+			podTemplateSpecHash, hashExists := podTmpl.ObjectMeta.Labels[extensions.DefaultDaemonSetUniqueLabelKey]
```
Can users override this key? If so, it's not safe to use it for picking up the desired pod template right?

lukaszo replied:
It's used in the same way as in deployments

kargakis replied:
In Deployments: we list all the replica sets for a deployment, [iterate over all replica sets](https://github.com/kubernetes/kubernetes/blob/f74fbfc04497b723c76b1e5baf15de550efc64dc/pkg/controller/deployment/util/deployment_util.go#L502), and [compare the pod template of the deployment with each rs pod template](https://github.com/kubernetes/kubernetes/blob/f74fbfc04497b723c76b1e5baf15de550efc64dc/pkg/controller/deployment/util/deployment_util.go#L497). If we find a match, that's our new replica set, if not then that means there is a change inside the deployment pod template and we need to create a new replica set.

[1]

kargakis commented on pkg/kubectl/history.go:
```diff
@@ -26,6 +26,7 @@ import (
 	"k8s.io/kubernetes/pkg/api/unversioned"
```
Can you split the kubectl change into a separate commit?

kargakis commented on pkg/kubectl/history.go:
```diff
@@ -127,3 +130,71 @@ func getChangeCause(obj runtime.Object) string {
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve DaemonSet %s: %v", name, err)
+	}
+	podTemplates, err := daemonutil.GetAllPodTemplates(daemon, h.c)
```
I would prefer to see using the pod template client directly instead of creating wrappers around it. We had been doing this in OpenShift deployments for some time and it was really painful to maintain.

kargakis commented on pkg/apis/extensions/types.go:
```diff
@@ -349,31 +340,35 @@ type DaemonSetSpec struct {
+	Paused bool `json:"paused,omitempty"`
+
+	// Indicates that the DaemonSet will automaticly update pods when DaemonSet Spec was changed.
+	AutoUpdate bool `json:"autoUpdate,omitempty"`
```
How about instead of a separate field, we default Paused to true? Paused for Deployments means exactly the opposite of AutoUpdate.

kargakis commented on pkg/apis/extensions/types.go:
```diff
@@ -349,31 +340,35 @@ type DaemonSetSpec struct {
 	// to daemon set pods to distinguish between old and new pod templates during
 	// DaemonSet update. See DaemonSetSpec's UniqueLabelKey field for more information.
-	DefaultDaemonSetUniqueLabelKey string = "daemonset.kubernetes.io/podTemplateHash"
+	DefaultDaemonSetUniqueLabelKey string = "pod-template-hash"
```
Shouldn't this be namespaced?

lukaszo replied:
It's the same as in Deployments https://github.com/lukaszo/kubernetes/blob/master/pkg/apis/extensions/types.go#L205

kargakis replied:
Hm, unfortunate. I would vote for having this namespaced for daemon sets and open an issue to change it for deployments. @bgrant0607 thoughts?

lukaszo replied:
Hmm, @bgrant0607 wanted it to be like now. See https://github.com/kubernetes/kubernetes/pull/19627/files#r51234701

kargakis commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -114,6 +114,33 @@ func ValidateDaemonSetStatusUpdate(ds, oldDS *extensions.DaemonSet) field.ErrorL
 	return allErrs
```
You are missing validation for observedGeneration and UpdatedNumberScheduled

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -444,27 +448,53 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
 
 		daemonPods, isRunning := nodeToDaemonPods[node.Name]
 
-		switch {
```
The switch is more readable imo

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -683,6 +740,48 @@ func (dsc *DaemonSetsController) nodeShouldRunDaemonPod(node *api.Node, ds *exte
+
+	needsUpdate := true
+	var podToRetain *api.Pod
+	newPodTemplateSpecHash := podutil.GetPodTemplateSpecHash(ds.Spec.Template)
```
Do we want to pick a different hashing algorithm? See https://github.com/kubernetes/kubernetes/issues/29735 for limitations in Adler. cc: @bgrant0607 @davidopp @erictune

lukaszo replied:
Oh. As I understand setting history limit is just a workaround and I should use new hashing algorithm?

kargakis replied:
If we are going to have this PR for 1.6, I would not bother with a different hashing algorithm and instead I would introduce `revisionHistoryLimit` for DaemonSets (which I guess we are going to have anyway) with a default set from start.

kargakis replied:
Eventually, I think we should use a different hashing algorithm (we are going to change it for Deployments too). @smarterclayton suggested fnv

kargakis commented on pkg/kubectl/rollout_status.go:
```diff
@@ -34,6 +34,8 @@ func StatusViewerFor(kind unversioned.GroupKind, c internalclientset.Interface)
 	switch kind {
```
Separate commit with kubectl changes - you could squash all kubectl changes into one commit if you want.

kargakis commented on pkg/apis/extensions/types.go:
```diff
@@ -305,6 +305,9 @@ type DaemonSetUpdateStrategyType string
 	RollingUpdateDaemonSetStrategyType DaemonSetUpdateStrategyType = "RollingUpdate"
+
+	// Dont touch daemon pods.
+	NoopDaemonSetStrategyType DaemonSetUpdateStrategyType = "Noop"
```
Ah, much better than a separate api field.

kargakis commented on pkg/controller/daemon/util/daemon_util.go:
```diff
@@ -147,6 +171,29 @@ func GetAllPodTemplates(daemon *extensions.DaemonSet, c clientset.Interface) (*a
+	// XXX:	deployment.Annotations = getSkippedAnnotations(deployment.Annotations)
+	daemon.Annotations = make(map[string]string)
+	for k, v := range rollbackToPT.Annotations {
+		// XXX:if !skipCopyAnnotation(k) {
```
Add these back and have an empty map of the annotations we want to skip for now.

lukaszo replied:
ok

kargakis commented on test/e2e/daemon_set.go:
```diff
@@ -263,6 +264,53 @@ var _ = framework.KubeDescribe("Daemon set [Serial]", func() {
 
 	})
+
+	It("should update pod when spec was updated and update strategy is rolling update", func() {
```
You also need a test for paused daemon sets, for rollbacks, and for the noop strategy. I would be fine to have them in a follow-up.

lukaszo replied:
ok

kargakis commented:
A couple of comments in general:
* since this PR is huge I would suggest splitting it in smaller scoped PRs: 1) API changes 2) rolling upgrade changes in the controller 3) rollbacks+rollout undo 4) rest of the kubectl changes.
* did we end up deciding what to do with paused? In the current code, you just update the status of the daemon set. I think we want something similar to what happens in Deployments ie. we still manage the replicas for the current revision, we just don't run upgrades.
* you will need more e2e tests than what you currently have (I left a comment about them in the code)
* you will need to track (or preferrably create respective issues for DS) a couple of follow-up issues to this: canaries (https://github.com/kubernetes/kubernetes/issues/11505), daemon set conditions+maybe perma-failed daemon sets (https://github.com/kubernetes/kubernetes/issues/14519), autoRollbacks (https://github.com/kubernetes/kubernetes/issues/23211), infant mortality detection (https://github.com/kubernetes/kubernetes/issues/18568) are the most important i think

kargakis commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -444,27 +454,53 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
+				sort.Sort(podByCreationTimestamp(daemonPods))
+				podToRetain, needsUpdate := dsc.needsUpdate(ds, daemonPods, numUnavailable, len(nodeList.Items))
+
+				if needsUpdate {
```
I would separate the update part from `manage`.

lukaszo replied:
Yeah, it looks messy now

kargakis commented:
cc: @mfojtik

derekwaynecarr commented:
@bgrant0607 -- we do not support quota on `PodTemplate`.  like @kargakis this is my first time seeing it being used in actual storage.  OpenShift prohibits the creation of `PodTemplate` via authorization rules today.

fyi: @liggitt for post 1.5 review cycles.

lukaszo commented:
@kargakis first of all, big thans for your review.

> since this PR is huge I would suggest splitting it in smaller scoped PRs:

Sure, I will prepare seprate PRs
>1) API changes

The question is which version are we targeting. If 1.5 then @bgrant0607 suggests to do everything in annotations. Then no api changes will be required. The question is which path to take?
> did we end up deciding what to do with paused? In the current code, you just update the status of the daemon set. I think we want something similar to what happens in Deployments ie. we still manage the replicas for the current revision, we just don't run upgrades.

It looks like I missread the code for pause in Deployment. I will fix the behavior.

> you will need more e2e tests than what you currently have (I left a comment about them in the code)

Will do

>  you will need to track (or preferrably create respective issues for DS) a couple of follow-up issues to this: canaries (#11505), daemon set conditions+maybe perma-failed daemon sets (#14519), autoRollbacks (#23211), infant mortality detection (#18568) are the most important i think

I will create new issues for daemon sets

lukaszo commented on pkg/apis/extensions/v1beta1/conversion.go:
```diff
@@ -41,6 +41,12 @@ func addConversionFuncs(scheme *runtime.Scheme) error {
 		Convert_v1beta1_DeploymentStrategy_To_extensions_DeploymentStrategy,
 		Convert_extensions_RollingUpdateDeployment_To_v1beta1_RollingUpdateDeployment,
 		Convert_v1beta1_RollingUpdateDeployment_To_extensions_RollingUpdateDeployment,
+		Convert_extensions_DaemonSetSpec_To_v1beta1_DaemonSetSpec,
```
Hmm, this is part from rebased  #19627. I didn't check if it still needed. I will check it.

lukaszo commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -626,6 +678,11 @@ func (dsc *DaemonSetsController) syncDaemonSet(key string) error {
 	if err != nil {
 		return fmt.Errorf("couldn't get key for object %#v: %v", ds, err)
 	}
+	if ds.Spec.Paused {
```
I've misread how it's working in deployment. I will fix it

kargakis commented:
> The question is which version are we targeting. If 1.5 then @bgrant0607 suggests to do everything in annotations. Then no api changes will be required. The question is which path to take?

I slightly prefer to postpone this for 1.6 and have all changes into the API, since I think it's already established in Deployments.

idvoretskyi commented:
@bgrant0607 any update on this item? Initially, we have expected to see this item in 1.5, but regarding @kargakis mention, we have to decide should we move it to the next milestone.

bgrant0607 commented:
@idvoretskyi @lukaszo I'm sorry, but we really have no time to review this in 1.5.

@kargakis The reason for starting with annotations was because I wanted to start this feature in alpha. The implementation is fairly high risk, and I didn't want the feature enabled by default in 1.5. Additionally, due to the lack of time for a careful review, there's also an increased risk of needing to change the behavior later.

I had originally suggested a client-side implementation, which would have been simpler and lower risk, or an external controller-based implementation, which would have facilitated independent testing of the implementation.

kargakis commented:
> I had originally suggested a client-side implementation, which would have been simpler and lower risk, or an external controller-based implementation, which would have facilitated independent testing of the implementation.

I am fine with a kubectl implementation for this in `rolling-update` before we consider baking into the API. Not sure if we can have it in the 1.5 timeframe at this point.

clhodapp commented:
Is it reasonable that `DaemonSet`s should take care of updates for themselves when `ReplicaSet`s have externalized update scripting to `Deployment`s? It seems _annoyingly_ asymmetric.

sttts commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -521,21 +527,18 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
 	glog.V(4).Infof("Nodes needing daemon pods for daemon set %s: %+v, creating %d", ds.Name, nodesNeedingDaemonPods, createDiff)
+	podtemplate, err := daemonutil.GetOrCreatePodTemplate(dsc.podTemplateController, ds, dsc.kubeClient)
+	if err != nil {
+		glog.Errorf("Couldn't get PodTemplate for daemon set %s: %v", ds.Name, err)
```
Why not return the error? If it's intended, it deserves a comment.

sttts commented on pkg/controller/daemon/util/daemon_util_test.go:
```diff
@@ -0,0 +1,220 @@
+		podTmpl, err := GetOrCreatePodTemplate(ptc, ds, fakeClient)
+		if err != nil {
+			t.Errorf("In test case %s, got unexpected error %v", test.test, err)
+		}
```
continue

sttts commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -444,27 +448,53 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
+					}
+					numUnavailable++
+					nodesNeedingDaemonPods = append(nodesNeedingDaemonPods, node.Name)
+				} else {
```
else if

sttts commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -444,27 +448,53 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
+						if podToRetain == nil {
+							podToRetain = daemonPods[0]
+						}
+						for i := 0; i < len(daemonPods); i++ {
```
`for i := 1.....` and no if clause

sttts commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -496,13 +526,24 @@ func (dsc *DaemonSetsController) manage(ds *extensions.DaemonSet) error {
-			if err := dsc.podControl.CreatePodsOnNode(nodesNeedingDaemonPods[ix], ds.Namespace, &ds.Spec.Template, ds); err != nil {
+
+			// Make a copy
+			template := ds.Spec.Template
```
This is not a copy. Use `DeepCopy`.

bgrant0607 commented:
External implementation:
https://github.com/Mirantis/k8s-daemonupgradecontroller/blob/master/README.md

bgrant0607 commented:
cc @piosz

janetkuo commented on test/e2e/daemon_set.go:
```diff
@@ -263,6 +264,53 @@ var _ = framework.KubeDescribe("Daemon set [Serial]", func() {
+
+		framework.Logf("Creating simple daemon set %s", dsName)
+		_, err := c.DaemonSets(ns).Create(&extensions.DaemonSet{
+			ObjectMeta: api.ObjectMeta{
```
Make this part a function for easier test setup

janetkuo commented on test/e2e/daemon_set.go:
```diff
@@ -374,3 +422,24 @@ func checkRunningOnAllNodes(f *framework.Framework, selector map[string]string)
+			podImage := pod.Spec.Containers[0].Image
+			if podImage != image {
+				err := fmt.Errorf("Wrong image for pod: %s. Expected: %s, got: %s", pod.Name, image, podImage)
+				return false, err
```
We shouldn't return error here. It gives up the wait loop immediately. We should instead return `nil` to wait for the image to match.

janetkuo commented:
Some more feedback:

1. `kubectl delete`:
    - `kubectl delete ds` should delete the DS's pods and podtemplates. Right now only pods are deleted.
    - Right now after `kubectl delete ds`, another podtemplate is created _(bug)_
2. Noop update strategy:
    - Right now `kubectl rollout status ds` hangs (it's still waiting for the DS to be udpated) _(bug)_
    - Right now a new podtemplate will be created upon DS udpate _(bug)_

janetkuo commented:
![](https://github.trello.services/images/mini-trello-icon.png) [DaemonSet update](https://trello.com/c/UKSJ2UjF/13-daemonset-update)

lukaszo commented:
@janetkuo the board is private

janetkuo commented:
> @janetkuo the board is private

That's my personal board. I'm not aware that it published comments here

janetkuo commented:
@lukaszo thanks for the good work. Would love to get this in 1.6. Are you going to continue working on this patch?

lukaszo commented:
@janetkuo yes, I would love to have it merged in 1.6. Unfortunatly it's not mine decision and I didn't get clear message if this is an acceptable approach. There is a sig-apps meeting on Monday where I hope it will be discussed. This patch is somehow connected with Statefull sets upgrades which also should be discussed there.
In the meantime I'm extracting some parts from here and sending as a separate patches. One is here #39157 and another is WIP here https://github.com/lukaszo/kubernetes/commit/54b44b39bf32f6b6176819bd8bc272f146a0308a after extracting I will rebase this PR.
For 1.5 I also created external controller which runs upgrades https://github.com/Mirantis/k8s-daemonupgradecontroller where I fixed some reviewer comments from here.

janetkuo commented:
@lukaszo sig-apps will discuss 1.6 planning in the next meeting (Jan. 9). Let's discuss this then

therc commented:
I can't make it for the meeting, but I had questions regarding high(er) availability during updates, both in the short and long term.

We can't guarantee perfect uptime, but we can reduce the time during which no pods of a given DaemonSet are running. A few alternatives:

- Some kind of rendezvous protocol like (IIRC) Envoy uses, where the old and new pods are briefly running together and coordinate handoff. This is like the 1% case. Very nice to have, but a long stretch.
- Reuse the idling proxy, when it will be available. (Can't look up the issue right now.)  When a pod's update starts, its connections are sent to a special proxy and/or kube-proxy, which stall the traffic until the new pod is ready to serve. Easier to stall with TCP than UDP (which e.g. DataDog's statsd uses).
- The nuclear option: drain the entire node, update pod, undrain.

Comments?

kargakis commented:
The simplest upgrade process I can think of mimics rolling deployments with
maxSurge=0 and maxUnavailable=1:  scale down old pod, scale up new pod,
wait for new pod to become ready, repeat for every node. The Envoy case
sounds like an advanced strategy and should be interesting to explore in
the future.

On Mon, Jan 9, 2017 at 5:55 PM, Rudi C <notifications@github.com> wrote:

> I can't make it for the meeting, but I had questions regarding high(er)
> availability during updates, both in the short and long term.
>
> We can't guarantee perfect uptime, but we can reduce the time during which
> no pods of a given DaemonSet are running. A few alternatives:
>
>    - Some kind of rendezvous protocol like (IIRC) Envoy uses, where the
>    old and new pods are briefly running together and coordinate handoff. This
>    is like the 1% case. Very nice to have, but a long stretch.
>    - Reuse the idling proxy, when it will be available. (Can't look up
>    the issue right now.) When a pod's update starts, its connections are sent
>    to a special proxy and/or kube-proxy, which stall the traffic until the new
>    pod is ready to serve. Easier to stall with TCP than UDP (which e.g.
>    DataDog's statsd uses).
>    - The nuclear option: drain the entire node, update pod, undrain.
>
> Comments?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31693#issuecomment-271339397>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf6jm3cssImyWZyzlP_0rhqMzz0v3ks5rQmaFgaJpZM4JwX1j>
> .
>

kargakis commented:
Btw we didn't come to a conclusion about this today in the absence of
Clayton and Brian and the next sig-apps meeting is in two weeks so we
should probably agree on this issue about how we should proceed. I think
the general desire is to provide a cli implementation first and then
promote to an api strategy similar to how `rolling-update` ended up in
Deployments.

On Tue, Jan 10, 2017 at 1:32 AM, Michail Kargakis <mkargaki@redhat.com>
wrote:

> The simplest upgrade process I can think of mimics rolling deployments
> with maxSurge=0 and maxUnavailable=1:  scale down old pod, scale up new
> pod, wait for new pod to become ready, repeat for every node. The Envoy
> case sounds like an advanced strategy and should be interesting to explore
> in the future.
>
> On Mon, Jan 9, 2017 at 5:55 PM, Rudi C <notifications@github.com> wrote:
>
>> I can't make it for the meeting, but I had questions regarding high(er)
>> availability during updates, both in the short and long term.
>>
>> We can't guarantee perfect uptime, but we can reduce the time during
>> which no pods of a given DaemonSet are running. A few alternatives:
>>
>>    - Some kind of rendezvous protocol like (IIRC) Envoy uses, where the
>>    old and new pods are briefly running together and coordinate handoff. This
>>    is like the 1% case. Very nice to have, but a long stretch.
>>    - Reuse the idling proxy, when it will be available. (Can't look up
>>    the issue right now.) When a pod's update starts, its connections are sent
>>    to a special proxy and/or kube-proxy, which stall the traffic until the new
>>    pod is ready to serve. Easier to stall with TCP than UDP (which e.g.
>>    DataDog's statsd uses).
>>    - The nuclear option: drain the entire node, update pod, undrain.
>>
>> Comments?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/31693#issuecomment-271339397>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ADuFf6jm3cssImyWZyzlP_0rhqMzz0v3ks5rQmaFgaJpZM4JwX1j>
>> .
>>
>
>

janetkuo commented:
@lukaszo I discussed with @bgrant0607 and @erictune in person and we think it'll be easier to discuss through a proposal and then implement based on that. I went ahead create a draft here https://github.com/kubernetes/community/pull/262, let's continue the discussions there

janetkuo commented:
@lukaszo now that the [proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/daemonset-update.md) is merged, we can start with:

1. First, implement OnDelete and RollingUpdate strategy without history.
2. Then, implement kubectl rollout status for easier observing DaemonSet update status.
3. Then the next step is to implement history (using PodTemplates) and its garbage collection policy (RevisionHistoryLimit). Then we add support for rollback, and implement all related kubectl commands (kubectl rollout history and kubectl rollout undo). 

We can do `1+2` in 1.6 and `3` in 1.7. Once we have rolling update implemented we can choose the more appropriate approaches (regarding history, rollback, drain, cordon, startup time, etc) based on community feedback. 

Would you like to update this PR or open new ones?

lukaszo commented:
@janetkuo currently I'm on vacation and I will be back on 30 January. To do 1 and 2 I just need to rebase the patch and remove some parts so it shouldn't take long. If you can not wait please sync with @pigmej and someone else from @kubernetes/mirantis can work on it.

janetkuo commented:
> @janetkuo currently I'm on vacation and I will be back on 30 January. To do 1 and 2 I just need to rebase the patch and remove some parts so it shouldn't take long. If you can not wait please sync with @pigmej and someone else from @kubernetes/mirantis can work on it.

Thanks. Would love to continue this earlier to get it merged in 1.6. @pigmej would you help find someone else to work on this?

pigmej commented:
@janetkuo what is the deadline (roughly) for submitting final patch for points 1 and 2?

janetkuo commented:
> what is the deadline (roughly) for submitting final patch for points 1 and 2?

@pigmej according to [1.6 proposed timeline](https://github.com/kubernetes/features/blob/master/release-1.6/release-1.6.md), it's Feb. 24, including tests for 1+2

pigmej commented:
@janetkuo doable by @lukaszo on time, after his vaccation :) Including tests and discussions ;)

janetkuo commented:
@lukaszo in your new PR, would you put the generated codes into a separate commit so that it's easier to rebase / review?

lukaszo commented:
@janetkuo sure, will do.

pires commented:
This is great work @janetkuo @lukaszo!

/cc'ing @philips @aaronlevy @luxas as this is most probably the solution for some self-hosted issues like upgrades for apiserver and checkpointer, and in the future an hypothetical _component upgrader_ tool for kubeadm, running on every node.

janetkuo commented:
@lukaszo is there an ETA on when the new PR will be ready for review?

lukaszo commented:
@janetkuo it should be ready on Friday/Monday

kargakis commented:
@lukaszo can't find the new PR and it seems you haven't opened it yet. Need any help with it? Based on the proposal that merged, it should be much simpler than what you have here, right?

lukaszo commented:
@kargakis I will create PR today(in few hours) and will describe everything there.

lukaszo commented:
New PR is here https://github.com/kubernetes/kubernetes/pull/41116
Closing this one.
